src/gt4py/next/otf/compiled_program.py

```diff
@@ -0,0 +1,275 @@
+        if self.static_params is None:
+            # this could also be done in `__call__` but would be an extra if in the fast path
+            self.static_params = ()
+        return tuple(self._param_index(p) for p in self.static_params)
+
+    def _compile_variant(
+        self,
+        static_args: dict[str, ScalarOrTupleOfScalars],
+        offset_provider: common.OffsetProviderType | common.OffsetProvider,
+    ) -> None:
+        assert self.static_params is not None
+        static_args = _sanitize_static_args(
+            self.definition_stage.definition.__name__, static_args, self.program_type
+        )
+
+        args = tuple(
+            arguments.StaticArg(value=static_args[name], type_=type_)
+            if name in self.static_params
+            else type_
+            for name, type_ in self.program_type.definition.pos_or_kw_args.items()
+        )
+
+        compile_time_args = arguments.CompileTimeArgs(
+            offset_provider=offset_provider,  # type:ignore[arg-type] # TODO(havogt): resolve OffsetProviderType vs OffsetProvider
+            column_axis=None,  # TODO(havogt): column_axis seems to a unused, even for programs with scans
+            args=args,
+            kwargs={},
+        )
+        key = _CompiledProgramsKey(
+            tuple(static_args[p] for p in self.static_params),
+            offset_provider
+            if common.is_offset_provider_type(offset_provider)
+            else common.offset_provider_to_type(offset_provider),
+        )
+        if key in self._compiled_programs:
+            raise ValueError(f"Program with key {key} already exists.")
+        self._compiled_programs[key] = _async_compilation_pool.submit(
+            self.backend.compile, self.definition_stage, compile_time_args=compile_time_args
+        )
+
+    def compile(
+        self,
+        offset_providers: list[common.OffsetProvider | common.OffsetProviderType],
+        **static_args: list[ScalarOrTupleOfScalars],
+    ) -> None:
+        """
+        Compiles the program for all combinations of static arguments and the given 'OffsetProviderType'.
+
+        Note: In case you want to compile for specific combinations of static arguments (instead
+        of the combinatoral), you can call compile multiples times.
+
+        Examples:
+            pool.compile(static_arg0=[0,1], static_arg1=[2,3], ...)
+                will compile for (0,2), (0,3), (1,2), (1,3)
+            pool.compile(static_arg0=[0], static_arg1=[2]).compile(static_arg=[1], static_arg1=[3])
+                will compile for (0,2), (1,3)
+        """
+        if self.static_params is not None:
+            if not sorted(self.static_params) == sorted(static_args.keys()):
+                raise ValueError(
+                    f"Static arguments must be the same for all compiled programs. Got {list(static_args.keys())}, expected {self.static_params}."
+                )
+        else:
+            self.static_params = tuple(static_args.keys())
```

Comment:
If we change the `self.static_params` type hint to `Collection`, we should also do:
```suggestion
            if set(self.static_params) != set(static_args.keys()):
                raise ValueError(
                    f"Static arguments must be the same for all compiled programs. Got {set(static_args.keys())}, expected {self.static_params}."
                )
        else:
            self.static_params = set(static_args.keys())
```